src/codegen.c

```diff
@@ -81,6 +81,17 @@ void codegen_emit(AST_expr *expr, int parent_numArgs, FILE *outputFile) {
 			fprintf(outputFile, "if_end%i:\n", label1);
 			break;
 
+        case When:
+			label1 = if_end_unique++;
+			codegen_emit(expr->body[0], parent_numArgs, outputFile);
+			fprintf(outputFile, "\tCPSE CRSh, falseReg\n\tCPSE falseReg, falseReg\n");
```

Comment:
CPSE is "ComPare, Skip (next instruction) if Equal". The JMP says "jump over all of the following instructions until the if_end label". The CRS register holds the result of the previous expression (which in this case is the 'when' condition). Therefore, these two instructions say "if the condition was equal to false then **don't** jump over these body expressions".

I don't think there's an inverse of CPSE, so do something like "CP CRSh, falseReg; BREQ if_end", which is "compare the result with false and branch to if_end if they were equal."

Reply:
Perfect; thanks! I knew there was something simpler. I'll see if I can spend a little time reading up on AVR before doing any more patches like this; do you have a resource you would recommend?

Reply:
I think the Atmel datasheets are pretty good, and [this](http://atmel-studio-doc.s3-website-us-east-1.amazonaws.com/webhelp/GUID-0B644D8F-67E7-49E6-82C9-1B2B9ABE6A0D-en-US-1/index.html) is my go-to reference for the instruction set. I can't think of any other resources in particular. When I was starting out I found it very useful to write simple C programs and compile to assembly (`avr-gcc -S`) or compile fully then disassemble with `avr-objdump` to inspect the result.